Category Import fixes

rbrown opened the pull request:
This has the compatibility fixes I commented on, sets a default category_path_seperator "/", and reverses the parameters for explodeEscaped because having the first parameter optional is deprecated.

rbrown committed:
Add a default category separator.

Without this explodeEsacped can be called without glue.

rbrown committed:
Use \DateTime again in Import\Category

The refactor changed this code to use  Magento\Framework\Stdlib\DateTime,
but it's constructor isn't the same as \DateTime, so I reverted it.

\DateTime->new accepting null is deprecated, so I also fixed that warning

rbrown committed:
Fix importing boolean attributes for categories

Import/Category->indexValueAttributes holds a list of attributes that
use the value of a select, rather than the label to update the
database. This doesn't match with the validation code, which uses
Magento\ImportExport\Model\Import::getAttributeType to

sprankhub commented:
@avstudnitz, do you think we can merge this?

sprankhub commented:
@rbrown, would you mind updating your PR and sending it to the current `develop` branch?

rbrown commented:
Hi @sprankhub I'm not sure I've done this right now. There should be three commits, 1 to add a default category separator, with this I didn't need to reorder the method parameters, 2. Use the right DateTime class and 3. A fix for importing boolean attributes.

sprankhub commented on Model/Import/Category.php:
```diff
@@ -117,9 +117,6 @@ class Category extends \Magento\ImportExport\Model\Import\AbstractEntity
      */
     private ?string $entityTable = null;
 
-    /**
```
Why don't we keep this?

sprankhub commented on Model/Import/Category.php:
```diff
@@ -335,7 +332,13 @@ public function getAttributeOptions(AbstractAttribute $attribute): array
+            $index = 'label';
+            if (
+                in_array($attribute->getAttributeCode(), $this->indexValueAttributes) ||
+                $attribute->getSourceModel() == "Magento\Eav\Model\Entity\Attribute\Source\Boolean"
```
Please use `::class` syntax and also import the class to match the general code style.

sprankhub commented on Model/Import/Category.php:
```diff
@@ -954,13 +957,13 @@ private function saveCategories(): self
                         CategoryInterface::KEY_PARENT_ID => $parentCategory['entity_id'],
                         CategoryInterface::KEY_LEVEL => $parentCategory[CategoryInterface::KEY_LEVEL] + 1,
-                        CategoryInterface::KEY_CREATED_AT => (new DateTime($time))
+                        CategoryInterface::KEY_CREATED_AT => (new \DateTime($time))
```
To match the general code style, let's import this and also remove the wrong use statement `Magento\Framework\Stdlib\DateTime`.

rbrown replied:
We're actually using both, \DateTime to initialise from a string and Magento\Framework\Stdlib\DateTime for DATETIME_PHP_FORMAT

sprankhub commented on etc/config.xml:
```diff
@@ -14,6 +14,7 @@
                 <validation_strategy>validation-stop-on-errors</validation_strategy>
                 <allowed_error_count>0</allowed_error_count>
                 <ignore_duplicates>0</ignore_duplicates>
+                <category_path_seperator>/</category_path_seperator>
```
I do not think we should do this, because this is a backwards-incompatible change. If people did not set this config value in their current systems, this will change it and possibly breaks imports.

rbrown replied:
Without a default, and without it being set in the config, you can't import. I think I had assumed the separator would be /. 

I think the error getting thrown was 

> In Import.php line 273:
> 
>   Please enter a correct entity model.

And I had to work backwards from that, and ended up seeing split getting called with no separator. I guess we could call it a documentation error?

I don't think adding the default will break anything though, as it wasn't working before, and anyone who sets a value will be ignoring the default.

sprankhub replied:
Pretty sure it only breaks with PHP 8. With PHP 7, it would silently be ignored. And if we now configure it with a slash, people importing a category called "Shirts / T-Shirts" would end up with a "Shirts" parent category and a "T-Shirts" child category, whereas previously, they just got one category called "Shirts / T-Shirts". So I still think this needs to be reverted in order to not break existing imports :-/

rbrown replied:
No problem. I added a revert for that commit

sprankhub commented:
Thanks a lot, @rbrown! I hope my comments make sense to you :)

rbrown committed:
restore comment

rbrown committed:
Fix $config being used before it is initialised

rbrown committed:
Style updates

https://github.com/firegento/FireGento_FastSimpleImport2/pull/121/files/89aa66e7bacef86511b148297ee1cae833cea457#r1002145549

rbrown committed:
Revert "Add a default category separator."

This reverts commit db3d2ceb354b1d55b14710f6c3722419715527b1.

sprankhub commented:
Thanks, @rbrown!
